README
GENERAL:

NOTE: pbxbuild is deprecated and is no longer maintained.  Please use
      the buildtool + Xcode framework instead.
 
This tool converts XCode(tm) projects into GNUmakefiles and builds them.
Those GNUmakefiles may be the basis for further user modifications
in order to make up for the platform differences of GNUstep and OS X.
Currently OSX Version 3.9 and 4.2 project files are supported.

It works fairly well for small and medium sized projects.
However, this tool has no magic superpowers, especially:
	- It only processes a part of the project file necessary to
	  create generic GNUmakefiles
	- It cannot know about platform issues (e.g. whether there are
	  modifications needed in order to compile the OS X project
	  correctly on GNUstep)

BUILD / INSTALL:
Go to the project directory and type
make
make install

note: cmdline.c and cmdline.h are generated out of the file
optionsparser.ggo using GNU gengetopt
(www.gnu.org/software/gengetopt/gengetopt.html), so please
dont modify them directly.

USE:
getting help:                   pbxbuild -h

building a project 
in the current directory:       pbxbuild

generating the makefiles only:  pbxbuild -g

OUTPUT:
A directory pbxbuild is created. It contains an aggregate GNUmakefile and a
directory for every Target in the project with subproject
GNUmakefiles. The target directories have the source tree symlinked
into it.

CAUTION: 
Every invocation of pbxbuild removes the previous pbxbuild tree 
without warning!

Bug reports, comments and wishes welcome,
please send to hansfbaier (AT) googlemail (dot) com



ChangeLog
2014-04-28 01:39-EDT Gregory John Casamento <greg.casamento@gmail.com>

	* README: Note deprecation of pbxbuild in comment.

2012-03-08 18:56-EST Gregory John Casamento <greg.casamento@gmail.com>

	* Undo previous reversion.

2012-03-08 18:56-EST Gregory John Casamento <greg.casamento@gmail.com>

	* cmdline.c
	* cmdline.h
	* GNUmakefile
	* GNUmakefile.preamble
	* main.m
	* PBDevelopmentHelper.m
	* PBMakefileGenerator.h
	* PBMakefileGenerator.m
	* PBPbxGroup.h
	* PBPbxGroup.m
	* PBPbxNativeTarget.h
	* PBPbxNativeTarget.m
	* PBPbxProject.h
	* PBPbxProject.m
	* PBProjectGenerator.h
	* PBProjectGenerator.m: Reverting back to revision 29000.  pbxbuild is
	broken on the current revision for most purposes.  Also pbxbuild is
	being superceded by buildtool and the xcode/xclib framework which does
	not generate makefiles.

2011-08-13 Fred Kiefer <FredKiefer@gmx.de>

	* ChangeLog: Re-add change logs deleted by Greg's revert.
	* PBPbxNativeTarget.m: Ignore missing INFOPLIST_FILE entries.
	* PBMakefileGenerator.m (-insertCFlagsForTarget:inMakefile:): Keep
	CPPFLAGS on one line.

2011-08-02 Gregory Casamento <greg.casamento@gmail.com>

	* *.[hm]: Undid rollback

2011-07-28 Gregory Casamento <greg.casamento@gmail.com>

	* *.[hm]: Rollback to 29000.

2010-12-28 Fred Kiefer <FredKiefer@gmx.de>

 	* PBPbxNativeTarget.h,
	* PBPbxNativeTarget.m,
	* PBPbxProject.m,
	* main.m: Add some support for Jaguar projects.

2010-01-26 Eric Wasylishen <ewasylishen@gmail.com>

 	* PBPbxNativeTarget.h: Rename PBX_VERSION_SNOWLEOPARD to
	PBX_VERSION_SNOWLEOPARD_XCODE_3_1 and add
	PBX_VERSION_SNOWLEOPARD_XCODE_3_2.
	* PBPbxNativeTarget.m: Use the Tiger/Panther method of getting
	the Info.plist path on Tiger and later.
	Modify the search for "buildSettings" on XCODE_3_1 and XCODE_3_2
	to also check the build configuration list (same as Tiger and Leopard)

2009-12-15 04:29-WIT Hans Baier <hansfbaier@googlemail.com>

	* PBPbxMakefileGenerator.m: fix crash bug: double release

2009-12-01 17:55-EST Gregory John Casamento <greg.casamento@gmail.com>

	* PBPbxNativeTarget.m: Replace @try/@catch with NS_DURING
	and NS_HANDLER/ENDHANDLER

2009-11-20 Hans Baier <hansfbaier@goolemail.com>

	* applied Joe Matiellos patch after some style cleanup

2009-09-02 Riccardo Mottola <rmottola@users.sf.net>
	* PBPbxNativeTarget.[hm]: added snow leopard pbx constant

2009-08-16 09:36-EDT Gregory John Casamento <greg.casamento@gmail.com>

	* PBPbxNativeTarget.m: Change to allow compilation of dynamic libs.

2009-03-09 14:55-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* PBPbxNativeTarget.m: Correct issue with building target.
	* PBProjectGenerator.m: Temporarily remove code which generates the
	PC.project contents.

2009-03-06 22:01-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m: Correct C++ compilation issue.
	* PBPbxNativeTarget.m: Allow adding multiple times to the
	source arrays.

2009-03-02 02:23-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m: Move include for common.make.

2009-02-28 00:46-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m: Create scripts in scripts directory under
	pbxbuild so that they can be executed from there.
	* PBPbxNativeTarget.m: Change to dictionary.
	* PBProjectGenerator.m: Cleanup.

2009-02-27 10:32-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m: add scripts to before-all::
	* PBPbxNativeTarget.[hm]: Changes to store script information.

2009-02-26 23:40-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* main.m: Cleanup
	* PBMakefileGenerator.m: Add return files.
	* PBPbxNativeTarget.m: Fixes for static libraries.

2009-02-24 16:27-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* main.m: Compile all static if one project is a static lib.
	* PBMakefileGenerator.m: Final changes to support static libraries.
	* PBPbxNativeTarget.m: Removed unecessary autorelease/retain calls.

2009-02-24 02:40-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* PBPbxNativeTarget.h
	* PBPbxNativeTarget.m: Changes to support leapard style projects.

2009-02-23 21:19-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m
	* PBPbxNativeTarget.h
	* PBPbxNativeTarget.m: Changes to handle static libraries.

2009-01-02 Riccardo Mottola <rmottola@users.sf.net>

	* main.m
	* PBPbxNativeTarget.m, PBPbxProject.m, PBDevelopmentHelper.m,
	PBMakefileGenerator.m: Removed c-99isms

2008-11-12 01:49-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* main.m: Copy headers only from target directories.

2008-11-10 19:14-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* main.m: Copy all headers since many xcode projects
	don't include the headers in the project itself.
	Also copy the resources.
	* PBMakefileGenerator.m: Minor cleanup

2008-11-02 21:45-EST Gregory John Casamento <greg_casamento@yahoo.com>

	* cmdline.c
	* cmdline.h
	* main.m
	* optionsparser.ggo: Add option to use symbolic links instead
	of copying the files.

2008-10-26 21:00-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* main.m: Copy only those files needed for each target instead
	of linking ALL of the files in each directory.

2008-10-24 21:44-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m: Remove warning.
	* PBProjectGenerator.m: Fix issue adding nil to dictionary

2008-10-16 02:03-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* PBProjectGenerator.h
	* PBProjectGenerator.m: Added class to handle generation of the
	PC.project file.

2008-10-16 02:02-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* GNUmakefile
	* main.m
	* PBMakefileGenerator.m: Added code to generate PC.project file.

2008-10-12 21:46-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* PBMakefileGenerator.m: Add QuartzCore to the list of excluded
	libraries.

2008-10-12 21:43-EDT Gregory John Casamento <greg_casamento@yahoo.com>

	* main.m

main.m
/*
   Project: pbxbuild

   Copyright (C) 2006 Free Software Foundation

   Author: Hans Baier

   Created: 2006-08-09 04:23:23 +0200 by jack

   This application is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.

   This application is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU General Public
   License along with this library; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111 USA.
*/

#include "cmdline.h"
#include <stdlib.h>

#include <Foundation/Foundation.h>
#include "PBMakefileGenerator.h"
#include "PBProjectGenerator.h"
#include "PBPbxProject.h"

NSString *
findProjectFilename(NSArray *projectDirEntries);

/**
 * <p>This tools builds XCode-Projects by converting the project files into 
 * a directory hierarchy under <code>pbxbuild</code>, each target in the
 * original project is represented by a GNUmakefile subproject</p>
 */
int
main(int argc, const char *argv[], char *env[])
{
  PBPbxProject               *project;
  NSString                   *projectFilename;
  NSString                   *projectDir;
  NSArray                    *projectDirEntries;
  NSString                   *projectMakefile;
  NSString                   *projectFile;
  PBPbxNativeTarget          *target;
  PBMakefileGenerator        *makefileGenerator;
  PBProjectGenerator         *pcGenerator;
  struct gengetopt_args_info args_info;
  NSEnumerator               *e;
  NSFileManager              *fileManager;
  NSString                   *pbxbuildDir;
  NSString                   *pcfile;

  NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
  fileManager = [NSFileManager defaultManager];

  /* let's call our cmdline parser */
  if (cmdline_parser (argc, argv, &args_info) != 0)
    {
      cmdline_parser_print_help();
      [pool release];
      exit(EXIT_FAILURE);
    }

  if (args_info.help_given)
    {
      cmdline_parser_print_help();
      [pool release];
      exit(EXIT_SUCCESS);
    }

  if (args_info.version_given)
    {
      cmdline_parser_print_version();
    }

  if (args_info.debug_given)
    {
      #ifdef GNUSTEP
      NSMutableSet *debugSet;
      [[NSProcessInfo processInfo] setDebugLoggingEnabled: YES];
      debugSet = [[NSProcessInfo processInfo] debugSet];
      [debugSet addObject: @"dflt"];
      #endif
    }

  // get the direntries of the current directory
  projectDir        = [fileManager currentDirectoryPath];
  projectDirEntries = [fileManager directoryContentsAtPath: projectDir];

  // get the project filename
  if (args_info.project_given) 
    projectFilename = [[NSString stringWithCString: args_info.project_arg]
			stringByAppendingPathComponent: @"project.pbxproj"];
  else
    projectFilename = findProjectFilename(projectDirEntries);

  if (projectFilename == nil || [@"" isEqual: projectFilename])
    {
      NSLog(@"No project (.xcode, .xcodeproj) found in current directory!");
      RELEASE(pool);
      exit(EXIT_FAILURE);
    }

  // initialize project model an makefile generator
  project           = [[PBPbxProject alloc] initWithFile: projectFilename];
  makefileGenerator = [[PBMakefileGenerator alloc] initWithProject: project];
  pcGenerator       = [[PBProjectGenerator alloc] initWithProject: project];

  // create (overwrite) pbxbuild directory for builds
  pbxbuildDir = [[fileManager currentDirectoryPath] 
		   stringByAppendingPathComponent: @"pbxbuild"];

  if ([fileManager fileExistsAtPath: pbxbuildDir])
    {
      NSLog(@"Removing old build dir...\n");
      [fileManager removeFileAtPath: pbxbuildDir handler: nil];
    }

  [fileManager createDirectoryAtPath: pbxbuildDir attributes: nil];

  // create project makefile and PC.project file.
  projectMakefile = [makefileGenerator generateProjectMakefile];
  NSDebugLog(@"Project Makefile:\n%@\n", projectMakefile);
  [projectMakefile writeToFile: 
		     [pbxbuildDir 
		       stringByAppendingPathComponent: @"GNUmakefile"]
		   atomically: YES];

  projectFile = [pcGenerator generateProjectFile];
  NSDebugLog(@"Project file:\n%@\n", projectFile);
  [projectFile writeToFile: 
		     [pbxbuildDir 
		       stringByAppendingPathComponent: @"PC.project"]
		   atomically: YES];

  // generate subprojects
  e = [[project targets] objectEnumerator];
  while ( (target = [e nextObject]) )
    {
      NSEnumerator *f = [projectDirEntries objectEnumerator];
      NSString   *projectDirEntry;
      NSString   *makefile;
      NSString   *newTName = [[target targetName] stringByReplacingString: @" "
						  withString: @"_"];

      NSString   *targetDir = 
	[pbxbuildDir stringByAppendingPathComponent:
		       [newTName
			 stringByAppendingPathExtension: [target extension]]];

      [fileManager createDirectoryAtPath: targetDir attributes: nil];
      
      // link all dir entries of the project directory into the target dir
      while ( (projectDirEntry = [f nextObject]) ) 
	{
	  NSString *destination = 
	    [targetDir stringByAppendingPathComponent: projectDirEntry];

	  // skip existing GNUmakefiles
	  if ([projectDirEntry hasPrefix: @"GNUmakefile"])
	    continue;
					       
#ifdef __MINGW32__
	  {
	    NSString *source = 
	      [projectDir stringByAppendingPathComponent: projectDirEntry];
	    NSDebugLog(@"Copying from '%@' to '%@'", 
		  source,
		  destination);
	    [fileManager copyPath: source
			 toPath: destination
			 handler: nil];
	  }
#else
	  {
	    NSString *source = 
	      [@"../../"  stringByAppendingPathComponent: projectDirEntry];
	    NSDebugLog(@"Creating symbolic link from '%@' to '%@'", 
		       source,
		       destination);
	    [fileManager createSymbolicLinkAtPath: destination
			 pathContent: source];
	  }
#endif
	}

      // generate and write makefile
      makefile = [makefileGenerator generateMakefileForTarget: target];
      NSDebugLog(@"Makefile for target: '%@':%@\n", 
		 [target targetName], 
		 makefile );
      [makefile writeToFile: 
		  [targetDir stringByAppendingPathComponent: @"GNUmakefile"]
		atomically: YES];

      pcfile = [pcGenerator generateProjectForTarget: target];
      NSDebugLog(@"Project file for target: '%@':%@\n", 
		 [target targetName], 
		 makefile );
      [pcfile writeToFile: 
		[targetDir stringByAppendingPathComponent: @"PC.project"]
	      atomically: YES];

      // create link to Info.plist file
      NSString * infoPlistTargetPath = [targetDir stringByAppendingPathComponent: 
                                                      [NSString stringWithFormat: @"%@Info.plist", [target targetName]]];
      if ([target infoPlistFile] != nil)
        {
          [fileManager 
		   copyPath:	[projectDir stringByAppendingPathComponent: [target infoPlistFile]]
                     toPath:		infoPlistTargetPath
                    handler:		nil];
          
          //Fix XCode variables
          NSString *plistString = [NSString stringWithContentsOfFile:infoPlistTargetPath];
	
          plistString = [plistString stringByReplacingString: @"${EXECUTABLE_NAME}" withString:[target targetName]];
          plistString = [plistString stringByReplacingString: @"${PRODUCT_NAME:identifier}" withString:[target targetName]];
          [plistString writeToFile:infoPlistTargetPath atomically:YES];
        }
      else 
        {
          [[target infoPlist] writeToFile:infoPlistTargetPath atomically: YES];		
        }
    }

  // if user wants to generate makefile only, exit here
  if(args_info.generate_makefile_only_given) 
    {
      AUTORELEASE(project);
      AUTORELEASE(makefileGenerator);
      AUTORELEASE(pcGenerator);
      RELEASE(pool);
      exit(EXIT_SUCCESS);
    }

  // finally changedir to the pbxbuild directory and run make
  [fileManager changeCurrentDirectoryPath: @"pbxbuild"];

  int exitstatus = system("make -k");

  AUTORELEASE(project);
  AUTORELEASE(pcGenerator);
  AUTORELEASE(makefileGenerator);
  RELEASE(pool);
  
  return EXIT_SUCCESS;
}

NSString *
findProjectFilename(NSArray *projectDirEntries)
{
  NSEnumerator *e = [projectDirEntries objectEnumerator];
  NSString     *fileName;

  while ((fileName = [e nextObject]))
    {
      if (   [[fileName pathExtension] isEqual: @"xcode"]
	  || [[fileName pathExtension] isEqual: @"xcodeproj"]
	  || [[fileName pathExtension] isEqual: @"pbproj"] )
	return [fileName stringByAppendingPathComponent: @"project.pbxproj"];
    }

  return nil;
}
